### DISCUSSION TITLE
Improved Yolov5 Genetic Algorithm

### DISCUSSION THREAD

**Discussion by ShAmoNiA** [ultralytics/yolov5#11124] [Ideas]
Greeting,

So, long time age I started to use evolotion part in yolo; Howerver, it just contatins mutation part. 
That's why I said to myself, why shouldn't we have a more perfect model?
For this reason, I prepared the following code, which, in addition to mutation , also contains cross-over, which we usually use in our genetic algorithms. I hope it will be useful for you too.

## hyperparameters, the initial population, and the evolution loop:
```
import random
import numpy as np

# Define hyperparameters to be optimized
hyperparameters = {
    'learning_rate': np.arange(0.001, 0.01, 0.001),
    'batch_size': [16, 32, 64],
    'num_epochs': [10, 20, 30],
    'num_layers': [3, 4, 5],
    'image_size': [416, 512, 608]
}

# Define the population size and number of generations
POPULATION_SIZE = 10
NUM_GENERATIONS = 50

# Define the mutation rate
MUTATION_RATE = 0.1

# Define the initial population
population = []
for i in range(POPULATION_SIZE):
    individual = []
    for param in hyperparameters.values():
        individual.append(random.choice(param))
    population.append(individual)

# Evolution loop
for generation in range(NUM_GENERATIONS):
    # Calculate the fitness scores for each individual in the population
    fitness_scores = []
    for individual in population:
        fitness_score = evaluate_fitness(individual)
        fitness_scores.append(fitness_score)
    
    # Select the parents for the next generation using tournament selection
    parents = []
    for i in range(POPULATION_SIZE):
        tournament = random.sample(range(POPULATION_SIZE), 5)
        fitness_tournament = [fitness_scores[j] for j in tournament]
        winner = tournament[np.argmax(fitness_tournament)]
        parents.append(population[winner])
    
    # Create the next generation by performing crossover and mutation on the parents
    new_population = []
    for i in range(POPULATION_SIZE):
        parent1, parent2 = random.sample(parents, 2)
        child = []
        for j in range(len(parent1)):
            if random.random() < 0.5:
                child.append(parent1[j])
            else:
                child.append(parent2[j])
        if random.random() < MUTATION_RATE:
            mutated_param_index = random.randint(0, len(hyperparameters)-1)
            mutated_param = list(hyperparameters.values())[mutated_param_index]
            child[mutated_param_index] = random.choice(mutated_param)
        new_population.append(child)
    
    # Replace the old population with the new one
    population = new_population
```
## evaluate_fitness:

```
from yolov5.train import train
from yolov5.utils.metrics import fitness

# Define the fitness function
def evaluate_fitness(individual):
    # Create the options and callback for training
    opt = {
        'epochs': individual[2],
        'batch_size': individual[1],
        'cfg': f'yolov5s{individual[3]}.yaml',
        'data': 'path/to/dataset.yaml',
        'hyp': 'path/to/hyp.yaml',
        'project': 'runs/optimization',
        'name': f'exp_{individual[3]}_{individual[4]}_{individual[1]}_{individual[2]}_{individual[0]}',
        'img_size': individual[4],
        'weights': 'yolov5s.pt',
        'device': 'cpu',
        'resume': False,
        'nosave': True,
        'notest': True,
        'multi_scale': False,
        'cache_images': False,
        'adam': False,
        'sync_bn': False,
        'local_rank': -1
    }
    callback = lambda metrics: fitness(metrics)

    # Train the model
    train(opt=opt, callback=callback)

    # Compute the fitness score
    return fitness[-1]['fscore']
```

I have plan to add this code in **train.py** too; I will post it here ASAP :)

**Comment by nopps07** [comment]
Amazing. I've been also trying to apply more advanced genetic algorithms. I'm looking forward to seeing more updates from your side.

**Comment by ShAmoNiA** [comment]
So, in the new version, I used the **geneticalgorithm2** library. 
In **train.py** simply, you need to just replace below code instead of **else** part in main function:
```
    else:

        opt.noval, opt.nosave, save_dir = True, False, Path(opt.save_dir)  # only val/save final epoch
        # ei = [isinstance(x, (int, float)) for x in hyp.values()]  # evolvable indices
        evolve_yaml, evolve_csv = save_dir / 'hyp_evolve.yaml', save_dir / 'evolve.csv'
        if opt.bucket:
            # download evolve.csv if exists
            subprocess.run([
                'gsutil',
                'cp',
                f'gs://{opt.bucket}/evolve.csv',
                str(evolve_csv),])
            
        varbound=np.array([[0,1],[0,15],[1.5,3.5],[3.5,5]])
        vartype=np.array(['real','int','real','real'])

        algorithm_param = {'max_num_iteration': 150,
                           'population_size':1000}

        model=ga(function=objective,dimension=4,variable_type_mixed=vartype,variable_boundaries=varbound,algorithm_parameters=algorithm_param)
        model.run(no_plot = False, time_limit_secs = time_to_seconds(days = 7, hours = 0,minutes = 0, seconds = 0))


        plot_evolve(evolve_csv)
        LOGGER.info(f'Hyperparameter evolution finished {opt.evolve} generations\n'
                    f"Results saved to {colorstr('bold', save_dir)}\n"
                    f'Usage example: $ python train.py --hyp {evolve_yaml}')
```

Just know some points about the above:
1. I changed opt.noval to False for improving GA speed. You can change it to what you want.
2. I need to declare my parameter range and type which I want to run GA for. For this part I declearded 4 var in something like this:
```
        varbound=np.array([[0,1],[0,15],[1.5,3.5],[3.5,5]])
        vartype=np.array(['real','int','real','real'])
```
You can change it for your approach in many ways you want.

Know I declared my objective function which was my goal to optimize it:
```
def objective(X):   
    
    global opt
    global device
    global callbacks
    
    iterations = 4
    
    with open('data/hyps/hyp-cage.yaml', errors='ignore') as f:
            hyp = yaml.safe_load(f)
    
    hyp["Applying_Cage"] = float(X[0])
    hyp["max_cageScaling"] = float(X[3])
    hyp["min_cageScaling"] = float(X[2])
    hyp["rotation"] = int(X[1])

    results_arr = []
    for _ in range(iterations):
        results_ = train(hyp.copy(), opt, device, callbacks=Callbacks())
        callbacks = Callbacks()
        keys = ('metrics/precision', 'metrics/recall', 'metrics/mAP_0.5', 'metrics/mAP_0.5:0.95', 'val/box_loss',
                'val/obj_loss', 'val/cls_loss')
    
        results, _, _ = validate.run(data = ****,weights = "runs/evolve/"+name+"/weights/best.pt")

        results_arr.append(results[2])
        print_mutation(keys, results, hyp.copy(), Path(opt.save_dir), opt.bucket)
    
    result = max(results_arr)
    
    return -(result*100)
```
My hyper parameters file was different from yolo default file and I opened it in this part as **hyp**. You can load yolo hyp.yaml to run GA for them.
Just know that, for every tuple of hyper parameters which I want to optimize our function based on, I train the yolo v5 model several times each time and you can change this iteration by changing **iterations** value. I guess 4 is appropriate for this purpose.

I will continue this idea for creating better GA in this discussion.

**Reply by PraveenMNaik** [reply]
Can you please let me know where to update the above code in yolov5

**Reply by ShAmoNiA** [reply]
Hi @PraveenMNaik 
Check this PR https://github.com/ultralytics/yolov5/pull/11348

**Reply by PraveenMNaik** [reply]
Great Sir, 
I am unable to test for my custom dataset. I am waiting eagerly to run, some problem associated with dependency. However, runs fine with COLAB, very slow.

**Reply by ShAmoNiA** [reply]
I will happy if I can help you @PraveenMNaik 
You can just copy from line 538 and replace with yolov5 repo to get rid of that dependency error.

```
 # Evolve hyperparameters (optional)
    else:
        # Hyperparameter evolution metadata (including this hyperparameter True-False, lower_limit, upper_limit)
        meta = {
        ....

```

**Reply by PraveenMNaik** [reply]
Sir, 
How many epochs to be given, you have given 2 epochs for 1000 generations, is it fair.

**Reply by PraveenMNaik** [reply]
> I will happy if I can help you @PraveenMNaik You can just copy from line 538 and replace with yolov5 repo to get rid of that dependency error.
> 
> ```
>  # Evolve hyperparameters (optional)
>     else:
>         # Hyperparameter evolution metadata (including this hyperparameter True-False, lower_limit, upper_limit)
>         meta = {
>         ....
> ```

Sir to your repo (  https://github.com/ShAmoNiA/yolov5/tree/patch-4 ) or yolov5 master repo ( https://github.com/ultralytics/yolov5 ) .

**Reply by ShAmoNiA** [reply]
Personally, I have been using yolov5m with 20 or 15 epochs for my model training.
You can use my ropo or you can cut the part (which I talked about it in the above comment) from my repo and paste it into yolov5 master branch.

**Reply by PraveenMNaik** [reply]
Sir, I have tried for my dataset but I couldn't get improvement in mAP@50....with dataset of 300 images pr class and having 3 classes, totally 900 images. any suggestion from your end Sir.

**Reply by ShAmoNiA** [reply]
Greeting @PraveenMNaik 
Would you please share your GA configs?

**Comment by PraveenMNaik** [comment]
Same .. I have not modified anything

**Reply by ShAmoNiA** [reply]
You needed to :)
That config was fundamental.
One of the most important parameters which you need to change it based on your dataset size in population size:
```
# GA configs
      pop_size = 2
```
If your dataset has more than 10000 image, I suggest you to set pop_size between 20 to 50; However for smaller dataset, we can use larger amount for pop_size. For your case I suggest you to use between 100 to 200. Just know that, the larger the size of this parameter, the longer it takes to complete evolve.
For instance, If you want to have A generations with pop_size = B, you need to train A*B models.
You can use smaller model like yolov5s or yolov5m for have faster training procedure.(don't use yolov5n)

**Reply by ShAmoNiA** [reply]
Another fact that you need to consider it, is which hyper parameters you want to tune.
As you know, a huge amount of our population are created with random values and  it will be hard to imagine we can have a set of random values which all of them have appropriate value for out training results.

Due to this attitude, I suggest you to split hyper parameters into some groups and in every evolution's procedure, just tune that specific group. Let's check below sample.
```
        meta = {
            'lr0': (False, 1e-5, 1e-1),  # initial learning rate (SGD=1E-2, Adam=1E-3)
            'lrf': (False, 0.01, 1.0),  # final OneCycleLR learning rate (lr0 * lrf)
            'momentum': (False, 0.6, 0.98),  # SGD momentum/Adam beta1
            'weight_decay': (False, 0.0, 0.001),  # optimizer weight decay
            'warmup_epochs': (False, 0.0, 5.0),  # warmup epochs (fractions ok)
            'warmup_momentum': (False, 0.0, 0.95),  # warmup initial momentum
            'warmup_bias_lr': (False, 0.0, 0.2),  # warmup initial bias lr
            'box': (False, 0.02, 0.2),  # box loss gain
            'cls': (False, 0.2, 4.0),  # cls loss gain
            'cls_pw': (False, 0.5, 2.0),  # cls BCELoss positive_weight
            'obj': (False, 0.2, 4.0),  # obj loss gain (scale with pixels)
            'obj_pw': (False, 0.5, 2.0),  # obj BCELoss positive_weight
            'iou_t': (False, 0.1, 0.7),  # IoU training threshold
            'anchor_t': (False, 2.0, 8.0),  # anchor-multiple threshold
            'anchors': (False, 2.0, 10.0),  # anchors per output grid (0 to ignore)
            'fl_gamma': (False, 0.0, 2.0),  # focal loss gamma (efficientDet default gamma=1.5)
            'hsv_h': (True, 0.0, 0.1),  # image HSV-Hue augmentation (fraction)
            'hsv_s': (True, 0.0, 0.9),  # image HSV-Saturation augmentation (fraction)
            'hsv_v': (True, 0.0, 0.9),  # image HSV-Value augmentation (fraction)
            'degrees': (True, 0.0, 45.0),  # image rotation (+/- deg)
            'translate': (True, 0.0, 0.9),  # image translation (+/- fraction)
            'scale': (True, 0.0, 0.9),  # image scale (+/- gain)
            'shear': (True, 0.0, 10.0),  # image shear (+/- deg)
            'perspective': (True, 0.0, 0.001),  # image perspective (+/- fraction), range 0-0.001
            'flipud': (True, 0.0, 1.0),  # image flip up-down (probability)
            'fliplr': (True, 0.0, 1.0),  # image flip left-right (probability)
            'mosaic': (True, 0.0, 1.0),  # image mixup (probability)
            'mixup': (True, 0.0, 1.0),  # image mixup (probability)
            'copy_paste': (True, 0.0, 1.0)}  # segment copy-paste (probability)
```
In this part we just plan to run our evolution code just on augmentations' hyper parameters(from hsv_h to copy_paste).
After reaching appropriate values for this parameters, we will set their flag to 'False' and set other hyper parameters flag like lr0, lrf and etc to 'True'

Just remember that, every hyper parameters which their flag is 'False' will be have static value and we will read their value form hyp.yaml file(likes hyp.scratch-low.yaml).

**Reply by PraveenMNaik** [reply]
> You needed to :) That config was fundamental. One of the most important parameters which you need to change it based on your dataset size in population size:
> 
> ```
> # GA configs
>       pop_size = 2
> ```
> 
> If your dataset has more than 10000 image, I suggest you to set pop_size between 20 to 50; However for smaller dataset, we can use larger amount for pop_size. For your case I suggest you to use between 100 to 200. Just know that, the larger the size of this parameter, the longer it takes to complete evolve. For instance, If you want to have A generations with pop_size = B, you need to train A*B models. You can use smaller model like yolov5s or yolov5m for have faster training procedure.(don't use yolov5n)

Sir, is there any particular reason to avoid yolov5n model., bcz for me yolov5n nad yolov5m gives more accurate result in normal run without considering  hyperparameter evolution.

**Reply by PraveenMNaik** [reply]
SIr, whats your imoplementation details. 
I am confused on the following. Are you using
1. Genetic algo having mutation and cross over. (I hope this)
2. Evolutionary technique?
3. Does it comes under Grid search, Random or Baysian or Its a seperate

**Reply by ShAmoNiA** [reply]
I forget to answer this reply.
This implantation conation mutation, cross over, elite child and tournament selection technique.

**Comment by PraveenMNaik** [comment]
Thank you sir for your response I will definitely give a try...ping back to
you Sir, if any doubts

**Reply by PraveenMNaik** [reply]
Sir, how did you run using coco dataset, bcz Yolov5 expects the images to be annotated in yolov5 format, however COCO dataset consists of no annotations?... have u done annotations from ur end?to run

**Reply by ShAmoNiA** [reply]
Hi again @PraveenMNaik 
You need to use ``` --data coco.yaml``` or ``` --data coco128.yaml``` when you try to run train.py.
Those flag will download coco and coco 128 dataset which have yolo v5 annotations' format.

**Comment by PraveenMNaik** [comment]
Sir, how about other configuration like
mutation_rate_min = 0.01
        mutation_rate_max = 0.5 (0.8 in case of master repo since they use mutation only)
        crossover_rate_min = 0.5
        crossover_rate_max = 1
        min_elite_size = 2
        max_elite_size = 5
        tournament_size_min = 2
        tournament_size_max = 10
for my dataset... can this be same as mentioned in your repo.

Sir which algorithm is used for selecting the child generated

**Reply by PraveenMNaik** [reply]
Sir I get below error:
 File "/home/yolov5-GA/train.py", line 691, in     tournament_fitness = [fitness_scores[j] for j in tournament_indices]
IndexError: list index out of range

**Reply by ShAmoNiA** [reply]
I think you said you are using COLAB for run this GA.
Would you please share it with me for checking this issue?

For other config, you can use default values.

**Comment by PraveenMNaik** [comment]
Sir  this I am running on local machine...colab is slow, hence switched to
this

**Reply by ShAmoNiA** [reply]
Can you clone and try again? 
Maybe you are using previous version of that PR.

**Reply by PraveenMNaik** [reply]
ok sir ,,, will give a try

**Reply by PraveenMNaik** [reply]
link please

**Reply by ShAmoNiA** [reply]
Just clone again from this PR which we talked about it.

**Reply by PraveenMNaik** [reply]
Hi Sir, 
I tried, what I observed was same things are getting repeated over a while, and process getting killed. 
Can you try once from your end , I will provide my small dataset. It would be great healing for me, bcz I am worried on this.

**Reply by ShAmoNiA** [reply]
Of course,
Please share your data set in google drive with me.
I will check it ASAP.

**Comment by PraveenMNaik** [comment]
Sir email please

**Reply by ShAmoNiA** [reply]
mousaviniashayan@gmail.com

**Comment by PraveenMNaik** [comment]
Sir,
I have shared all the files.
Regards,

**Comment by ShAmoNiA** [comment]
@PraveenMNaik 
check below link:
https://colab.research.google.com/drive/1pAghekBU0HVXzRVZEudUf0mACpRSAGlZ?usp=sharing

**Reply by PraveenMNaik** [reply]
Ok sir, thank you.

**Comment by ShAmoNiA** [comment]
For other questions you asked:
1. Yes, we try to optimize mAP0.5.
   **results** value is a tuple which the second element is mAP0.5. 
```
fitness_scores.append(results[2])
```
 

2. the tournament_size determines the number of individuals competing in each tournament during selection, influencing the balance between exploration and exploitation in the GA. In normal GA we just have constant value for tournament_size. In this implementation we used **Adaptive tournament**.We select tournament_size from **tournament_size_max** to **tournament_size_min** and this procedure apply by progressing in our generations. Typically for tournament selection, our tournament_size at least need to be 2; So we add max(2, tournament_size_min) to our code; In other hands our tournament_size must smaller than pop_size and due to that, we use min(tournament_size_max, pop_size). check below code(line):
```                
tournament_size = max(max(2, tournament_size_min), int(min(tournament_size_max, pop_size) - (generation / (opt.evolve / 10))))
```


3. opt.evolve is the number of generation which we want to have.

**Reply by PraveenMNaik** [reply]
sir, @ 2 why, "- (generation / (opt.evolve / 10))", 
why /10 is necessary , why not some other number other than 10.

And without these staement  (generation / (opt.evolve / 10), the algorithm cab be aid to be non-adaptive ...?

**Reply by ShAmoNiA** [reply]
Greeting again.
/10 is not necessary. You can replace with whatever you want.
Yes, without (generation / (opt.evolve / 10), our parameters will equal to *_size_max or *_rate_max.

**Comment by ExtReMLapin** [comment]
What settings would you advise for 1700 images (training set) (only one class, around 30-60k instances) , I have 3 A-100 (40gb) but I don't want it to run more than a week ?

**Reply by ShAmoNiA** [reply]
Greeting,
For getting appropriate results, you need to set poltation size at least = 50. By using higher value, the chance of getting better result will increase.
opt.evolve <= 20
poltation size <=50
must be suitable for you.

However if you are in time pressure, just use yolo GA.

**Reply by ExtReMLapin** [reply]
Alright, thanks for the fast answer, i'll stick with yolo GA for now.

**Comment by ShAmoNiA** [comment]
@PraveenMNaik did you asked something?

**Comment by PraveenMNaik** [comment]
Sir,
Welcome back, greetings of the day.

I tried your code for the configuration mentioned with same dataset.
However, my training results are in the range of 78 to 83% for ***@***.*** for
newGA and old GA.
When observed the results of your csv file, it has started from 50%+ to 80%
+ for ***@***.***,,, why is this difference.
Kindly let me know..... i used the dataset of same what u have used.

**Reply by ShAmoNiA** [reply]
I have changed a part of the code that is like this.
In the previous version that I had collected my csv files from, we were generating our population from completely random samples. Due to this attitude, our starting point had a lower mAP. In the current situation, the one you used, we created our populations from random samples and the yamls file in the hyp folder, together. This view helps us to have a better starting point for our GA.

**Reply by ShAmoNiA** [reply]
Just know that, most part of our population are creating from random sample.
Many times it may give us good results and many times it may not have good results.

**Reply by PraveenMNaik** [reply]
Do you have any idea about implementation idea about variable length genetic algorithm...and how to implement in your code...

**Reply by ShAmoNiA** [reply]
Greeting @PraveenMNaik 
Would please describe what means "variable length" in your idea?

**Comment by PraveenMNaik** [comment]
Hi Sir,
There is an issue in this repo...
Whenever we run code for 300 or any generations its doesn't conclude at that point instead keep evolving until we manually end it...this creates problem which doesn't produce result or outcome of the experiment performed... Kindly look into it....

**Reply by ShAmoNiA** [reply]
Hi @PraveenMNaik 
Sry for the late response.
Just one question, even with pop_size =1 and generation=1, the code will keep evolving?
The main point is that, the number of training sample is always equal to population size × generation.

**Reply by PraveenMNaik** [reply]
Yes

**Reply by ShAmoNiA** [reply]
it's weird.
I just ran it on [colab](https://colab.research.google.com/drive/1dXVWgjOQb5OVDgkhxcO8pGHTsd1fczn-?usp=sharing) and everything worked fine.
Be aware that the config was pop_size = 6.

Please check it and you are free to share your issue on colab with me.

**Comment by PraveenMNaik** [comment]
Ok ...hmmm.... for example if pop size=50 and in training session if i keep
evolve as 300 i expect it to complete it by 300 generations. But, in my
case it doesn't happen. Am I right?

**Reply by ShAmoNiA** [reply]
No! That is not correct.

Check this:
https://github.com/ultralytics/yolov5/discussions/11124#discussioncomment-5885001

If your evolve value is 300 and pop size is 50, you need to train 300*50 models.
So, your evolve.csv will have 15000 rows.

The definition of generation in GA which has crossover + mutation(like my GA) is kind of different from GA that only has mutation(like yolo GA).

In GAs which has crossover + mutation, because of having population, every generation will have population_size member; So, whenever we want to go to another generation, we will need to train our model in every member of our current population.
In GAs which only has mutation, we do not have any population!(seems like our population_size is 1); Based on this perspective, our train count will be equal to our generation size! 

If you want to use my GA with pop size=50 and want to only train 300 models, you are need to set evolve = (300/50).

Feel free to ask any qesution.

**Comment by PraveenMNaik** [comment]
Oh... that's informative.. .thanks

**Comment by PraveenMNaik** [comment]
By the way ....how about variable length genetic algorithm....u have any
idea of implementation...

**Reply by ShAmoNiA** [reply]
Not yet.
I will create PR whenever I mastered on these 2 papers.
[Genetic algorithm with variable length chromosomes for...](https://link.springer.com/article/10.1007/s11633-014-0870-x)
[Behavior Of Variable-length Genetic Algorithms Under...](https://stars.library.ucf.edu/cgi/viewcontent.cgi?article=4366&context=etd)
